### PR TITLE
Make automatic formatting and testing before a commit using git-hooks(husky)

### PR DESCRIPTION
**Pull request by sumanthyedoti**
- Formats the code according to guidelines using prettier before a commit with git-hooks using husky
- Also makes sure all tests are passed after formatting the code by running tests before a commit.

### CODE REVIEW

File: lib/astupdate.js [hunk 1]
```diff
@@ -1,122 +1,139 @@
...
+    if (patternParam.type === "ArrayExpression") {
+      patterncase.test = estemplate.literal("0");
       patterncase.test.sType = {
-        'type': 'array',
-        'elemTypes': {},
-        'commonType': 'needsInference',
-        'isHomogeneous': false
-      }
-      flag = true
+        type: "array",
```
Comment (angadgill92) [inline lib/astupdate.js]: Can this be modified to follow StandardJS? We prefer using single quotes and no semicolons (except where absolutely required)

Comment (sumanthyedoti) [reply lib/astupdate.js]: Changed to single quotes and no semicolons

**Comment by angadgill92** [comment]
Hey @sumanthyedoti thanks for taking the time to test this out. The git hook is going to be useful. I think we can use StandardJS instead of Prettier. You can find the docs [here](https://standardjs.com/). 

We could add the auto-format part to package.json itself, see [this](https://standardjs.com/#what-you-might-do-if-youre-clever).

Also, for common issues like tabs vs spaces, and how many spaces to use for soft tabs, you can add a .editorconfig file, most editors support it by default these days. The editors can read the config and automatically trim trailing spaces and add final newlines to files. Can read more about it [here](https://editorconfig.org/) 

For us, the .editorconfig should do this:

- Use spaces for indenting .js files
- Soft tabs (each tab is 2 spaces long)
- Trailing whitespaces should be trimmed, except in .md files
- Files should end with a newline at EOF

For the JS formatting we can use StandardJS instead of Prettier. 

We can keep Prettier for .md files for now (though I feel that's not really required).

**Comment by sumanthyedoti** [comment]
`standard` auto-formats with 2 spaces. Should `.editorconfig` be added?

**Comment by angadgill92** [comment]
> `standard` auto-formats with 2 spaces. Should `.editorconfig` be added?


Sure, `.editorconfig` will ensure that `standard` has lesser work to do. (It will only need to fix the syntax).

File: lib/gen-core.js [hunk 2]
```diff
@@ -0,0 +1,48 @@
+const fs = require('fs')
```
Comment (angadgill92) [inline lib/gen-core.js]: Can we have a separate PR for this change @sumanthyedoti ?

File: lib/include/core.js [hunk 3]
```diff
@@ -1,54 +1,58 @@
+/* eslint-disable-next-line no-unused-vars */
 class IOCore {
   constructor (ioFunc) {
-    this.then = cb => ioFunc((...args) => cb(...args));
-  };
+    this.then = (cbf) => ioFunc((...args) => cbf(...args))
```
Comment (angadgill92) [inline lib/include/core.js]: Let's keep `cb` instead of  `cbf`

File: test/combinationParsers.test.js [hunk 4]
```diff
@@ -67,8 +75,16 @@ const lambdaTest = lambda([id('a'), id('b')], binEx(id('a'), '+', id('b')))
...
+  binEx(id('a'), '+', id('b'))
+)
+testRunner(
+  lambdaCallParser(lambdaCallPass)[0],
+  lambdaCallTest,
+  'lambdaCallParser'
+)
 
 // const ioStmtPass = initObj(`a = putLine 'a'`)
 // const ioStmtTest = decl(id('a'), ioCall(id('putLine'), [str('a')]))
```
Comment (angadgill92) [inline test/combinationParsers.test.js:90]: Let's remove the commented out code.

File: test/combinationParsers.test.js [hunk 5]
```diff
@@ -21,9 +21,9 @@ const exprStmt = templ.expression
 
 /* Exported parsers from parser.js to test */
 const {
-  doBlockParser,
-  ioParser,
-  doFuncParser,
+  // doBlockParser,
```
Comment (angadgill92) [inline test/combinationParsers.test.js:24]: Let's remove the commented out code

File: lib/utilityFunctions.js [hunk 6]
```diff
@@ -24,27 +31,29 @@ const returnRest = (val, input, rest, field) => {
...
+  val.cursorLoc = { line: output.line, column: output.column }
   return [val, output]
 }
 
-const isEmptyObj = obj => {
-  for (let prop in obj) {
-    if (obj.hasOwnProperty(prop)) return false
+const isEmptyObj = (obj) => {
+  for (const prop in obj) {
+    if (Object.prototype.hasOwnProperty.call(obj, prop)) return false
```
Comment (angadgill92) [inline lib/utilityFunctions.js:44]: This is a nice catch, maybe talk about why this is preferred to `obj.hasOwnProperty` this Friday @sumanthyedoti

File: lib/typeInference.js [hunk 7]
```diff
@@ -9,36 +9,46 @@ const jsFunctions = require('./jsFunctions')
...
 
 const returnIdentifierType = (id, typeObj) => {
   if (isFunction(id.name, typeObj)) return getParamReturnTypes(id, typeObj)
   return typeObj[id.name]
 }
 
 const identifierType = (id, localTypeObj) => {
-  if (localTypeObj[id.name] !== undefined) return returnIdentifierType(id, localTypeObj)
-  if (globalTypesObj[id.name] !== undefined) return returnIdentifierType(id, globalTypesObj)
+  if (localTypeObj[id.name] !== undefined) { return returnIdentifierType(id, localTypeObj) }
```
Comment (angadgill92) [inline lib/typeInference.js:30]: Please change the eslint rule `curly` to `multi` instead of `all`

File: lib/parserObject.js [hunk 8]
```diff
@@ -14,56 +10,65 @@ const {
...
-    values.push(result[0])
-    rest = result[1]
-  }
-  return [values, rest]
-}
+parser.all =
+  (...parsers) =>
+    (src) => {
+      let rest = src
+      const values = []
```
Comment (angadgill92) [inline lib/parserObject.js:53]: Let's not change all `let` bindings to `const`, some of them are mutable within the function scope, and `const`  can create confusions.  @sumanthyedoti

File: lib/estemplate.js [hunk 9]
```diff
@@ -1,161 +1,186 @@
 const types = require('./operatorPrecedence')
 const notNull = require('./utilityFunctions').notNull
 const notUndefined = require('./utilityFunctions').notUndefined
-let estemplate = {}
+const estemplate = {}
```
Comment (angadgill92) [inline lib/estemplate.js:4]: Let's not change all `let` bindings to `const`

File: lib/clean.js [hunk 10]
```diff
@@ -1,4 +1,8 @@
 const argParser = require('yargs')
+
+const genCore = require('./gen-core')
 const cli = require('./cli')
+
 const argv = argParser(process.argv.slice(2)).argv
+genCore()
```
Comment (angadgill92) [inline lib/clean.js]: Let us make a separate PR for `genCore`

**Review by angadgill92 — CHANGES_REQUESTED** [review]
Hey @sumanthyedoti , I have brought the dev branch up to speed with master, please rebase before you update the PR 🙂

File: lib/gen-core.js (hunk 2 above)
Comment (sumanthyedoti) [reply lib/gen-core.js]: @angadgill92, I added this because the previous minification was causing errors. After rebase this would not be required I think as minification is removed. 

Do you think this decreases a little processing and recommend making another PR?

File: lib/include/core.js (hunk 3 above)
Comment (sumanthyedoti) [reply lib/include/core.js]: `standard` caused an error because of the name. Will try with an es-lint comment?

File: lib/typeInference.js (hunk 7 above)
Comment (sumanthyedoti) [reply lib/typeInference.js:30]: `standard` does not support custom rules it seems. Please [check here](https://standardjs.com/#i-disagree-with-rule-x-can-you-change-it)

File: lib/parserObject.js (hunk 8 above)
Comment (sumanthyedoti) [reply lib/parserObject.js:53]: This is auto-fix from `standard`. To bypass this, we need to add an es-lint comment

File: lib/estemplate.js (hunk 9 above)
Comment (sumanthyedoti) [reply lib/estemplate.js:4]: This is auto-fix from standard. To bypass this, we need to add an es-lint comment

File: lib/clean.js (hunk 10 above)
Comment (sumanthyedoti) [reply lib/clean.js]: @angadgill92 Do you think it is useful? After rebase this would not be required for the reason this is done.

In case you recommend doing this, now I minified using both `esprima`, and `escodegen`. Please suggest a good, simple minifyer.

File: lib/typeInference.js (hunk 7 above)
Comment (angadgill92) [reply lib/typeInference.js:30]: You can modify the `eslint` config in `package.json` to add this:
```json
{
  "eslint" : {
    "curly": "multi"
  }
}
```
There are options provided for some rules, you can check the docs for this specific rule [here](https://eslint.org/docs/rules/curly).

Similarly, there might be rule options that allow us not to change all `let` bindings to `const`.

File: lib/gen-core.js (hunk 2 above)
Comment (angadgill92) [reply lib/gen-core.js]: Yeah, I remember Santosh saying that minification is not required (mostly because it makes the generated code hard to read, and also minification can be handled by some other tool). 

Maybe we can have a CLI flag to minify this, but we should make it a separate PR. Let's also check with Santosh about this.
For now this can be removed.

File: lib/include/core.js (hunk 3 above)
Comment (angadgill92) [reply lib/include/core.js]: Could you share the error?

File: lib/typeInference.js (hunk 7 above)
Comment (sumanthyedoti) [reply lib/typeInference.js:30]: @angadgill92 I am sorry if I am missing something. I added the rules but saw no effect. 
`standard` says we can not customize rules. Please [See here](https://standardjs.com/#i-disagree-with-rule-x-can-you-change-it). For customizing rules, it says to use [`standardX`](https://github.com/standard/standardx).
Please correct me if I am missing anything.

File: lib/include/core.js (hunk 3 above)
Comment (sumanthyedoti) [reply lib/include/core.js]: It was this, from `standard` linter
```
Unexpected literal in error position of callback.
```

**Review by angadgill92 — APPROVED** [review]

File: lib/estemplate.js (hunk 9 above)
Comment (angadgill92) [reply lib/estemplate.js:4]: let it be then, let's add a `.eslintrc` in a future commit, to address this if required

File: lib/typeInference.js (hunk 7 above)
Comment (angadgill92) [reply lib/typeInference.js:30]: My bad, the config actually goes in `.eslintrc` let's leave it for now though, we can take it up later if required 👍🏼
